PyRDF/Node.py

```diff
@@ -171,3 +176,176 @@ def graph_prune(self):
+
+    """
+    def __init__(self, *args):
+        """
+        Creates a new RDataFrame instance for the given arguments.
+
+        Args:
+            *args (list): Variable length argument list to construct the
+                RDataFrame object.
+        """
+        super(HeadNode, self).__init__(None, None, *args)
+
+        args = list(args)  # Make args mutable
+        num_params = len(args)
+
+        for i in range(num_params):
+            # Convert Python list to ROOT CPP vector
+            if isinstance(args[i], list):
+                args[i] = self._get_vector_from_list(args[i])
+
+        try:
+            ROOT.RDataFrame(*args)  # Check if the args are correct
+        except TypeError as e:
+            msg = "Error creating the RDataFrame !"
+            rdf_exception = DataFrame.RDataFrameException(e, msg)
+            rdf_exception.__cause__ = None
+            # The above line is to supress the traceback of error 'e'
+            raise rdf_exception
+
+        self.args = args
+
+    def get_branches(self):
+        """Gets list of default branches if passed by the user."""
+        # ROOT Constructor:
+        # RDataFrame(TTree& tree, defaultBranches = {})
+        if len(self.args) == 2 and isinstance(self.args[0], ROOT.TTree):
+            return self.args[1]
+        # ROOT Constructors:
+        # RDataFrame(treeName, filenameglob, defaultBranches = {})
+        # RDataFrame(treename, filenames, defaultBranches = {})
+        # RDataFrame(treeName, dirPtr, defaultBranches = {})
+        if len(self.args) == 3:
+            return self.args[2]
+
+        return None
+
+    def _get_vector_from_list(self, arg):
+        """Converts a python list of strings to a vector."""
+        reqd_vec = ROOT.std.vector("string")()
+
+        for elem in arg:
+            reqd_vec.push_back(elem)
+
+        return reqd_vec
+
+    def get_num_entries(self):
+        """
+        Gets the number of entries in the given dataset.
+
+        Returns:
+            int: This is the computed number of entries in the input dataset.
+
+        """
+        first_arg = self.args[0]
+        if isinstance(first_arg, int):
+            # If there's only one argument
+            # which is an integer, return it.
+            return first_arg
+        elif isinstance(first_arg, ROOT.TTree):
+            # If the argument is a TTree or TChain,
+            # get the number of entries from it.
+            return first_arg.GetEntries()
+
+        second_arg = self.args[1]
+
+        # Construct a ROOT.TChain object
+        chain = ROOT.TChain(first_arg)
+
+        if isinstance(second_arg, str):
+            # If the second argument is a string
+            chain.Add(second_arg)
+        else:
+            # If the second argument is a list or vector
+            for fname in second_arg:
+                chain.Add(str(fname))
+
+        return chain.GetEntries()
+
+    def get_treename(self):
```

Comment:
What about this API? Is this intended to be invoked by the user?

Reply:
That's a good question. I think it has never been discussed.

Generally, all the `get_*` functions in `HeadNode` class are used internally in PyRDF to retrieve information that was presented by the user when constructing the dataframe, e.g. the name of the TTree or the filelist or the number of entries etc. . So I would say this should not be user facing

Reply:
Ok, then this class should not be returned to the user. Ideally, the user should see a distributed RDataFrame as having the same API (or a subset, at least) as the plain RDataFrame, but not introduce more methods (especially if there is something already equivalent in plain RDF).

Reply:
Ok let me understand this better. What I see when I construct a plain RDataFrame in PyROOT is
```
>>> import ROOT
>>> df = ROOT.RDataFrame(100)
>>> df.
df.Aggregate(              df.Define(                 df.GetColumnNames(         df.Histo2D(                df.Report(
df.Alias(                  df.DefineImpl(             df.GetColumnType(          df.Histo3D(                df.Snapshot(
df.AsNumpy(                df.DefineSlot(             df.GetDefinedColumnNames(  df.Max(                    df.SnapshotImpl(
df.Book(                   df.DefineSlotEntry(        df.GetFilterNames(         df.Mean(                   df.Stats(
df.Cache(                  df.Display(                df.GetNRuns(               df.Min(                    df.StdDev(
df.CacheImpl(              df.Fill(                   df.GetNSlots(              df.Profile1D(              df.Sum(
df.CheckAndFillDSColumns(  df.Filter(                 df.Graph(                  df.Profile2D(              df.Take(
df.Count(                  df.Foreach(                df.HasColumn(              df.Range(                  df.operator RInterface(
df.CreateAction(           df.ForeachSlot(            df.Histo1D(                df.Reduce(       
```

I'm not sure recreating this is possible, I would need to investigate. But here we're not giving a RDataFrame to the user, we're giving them SparkDataFrame or something similar. So generally it's not clear to me why the new distributed-aware dataframes should present the exact same API (or a subset of it)

Reply:
Especially for making this work with the backend itself it's necessary to have some methods or getter properties to retrieve said backend and work with that, isn't it?

Reply:
To clarify further, the API is gonna be there thanks to internal workings in PyRDF that connect the `Node` objects to their respective RDataFrame API callables, as was done so far in this package. Showing the same methods and attributes of the plain RDataFrame in the new `BackendDataFrame` classes is another thing though and that's where I'm in doubt

Reply:
I remember we actually discussed this in your presentation the other day, right? The idea is that the distributed RDFs look like plain RDFs, i.e. they have the same API (they can have less if it's not implemented). In the end we want the interface to be as similar as possible and the changes to the code to be minimal when switching backends.

Reply:
Ok so I need to hide somehow the fact that the [Spark]DataFrame instance can have some references to the [Spark] backend. I didn't think about it this way, need to rethink this through

Reply:
Latest commit has an implementation of this in the PyRDF.backend.Spark.SparkDataFrame class